students/luke/s03/mailroom.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+
+"""
+https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom.html#exercise-mailroom
+"""
+
+class donor:
+    """Donor class holds donor name and list of donations"""
+    def __init__(self, name, donations=None):
+        self.name = name
+        if donations is None:
+            self.donations = [ ]
+        else:
+            self.donations = donations
+
+
+def driver():
+    """ Main loop for mailroom """
+    done = False
+    while not done:
+        action = input(("Select an action and press return:\n"
+            "[S]end a thank-you (add users and donation entries)\n"
+            "[C]reate a report\n"
+            "[Q]uit\n"
+            "> "))
+        action = action.lower()
+        while not action in "scq": action = input("Bad option, reenter > ")
+
+        if action == "s":
+            add_donations()
+        elif action == "c":
+            create_report()
+        else:
+            done = True;
+    return
+
+
+def add_donations():
+    """ Add donation history for a(n optionally new) user """
+    done = False
+    while not done:
+        name = input("Enter donor name (or \"list\" for list): ")
+        if name == "list":
+            # list donor names
+            for d in donor_history: print(d.name)
+            continue
+        for thisdonor in donor_history:
+            if name == thisdonor.name:
+                break
+        if thisdonor == None:
+            thisdonor = donor(name)
+            donor_history.append(thisdonor)
+            print("Adding new donor: " + name)
+        moredonations = True
+        while moredonations:
+            value = input("Enter donation amount or -1 when finished: ")
```

Comment:
It might be less awkward to allow the user to just hit enter instead of -1. Then you can change your flow control to: 
if donation_amount == "": break

Reply:
Good point.  I was focused on validating input more than terminating the loop.  Fixed in next revision of s04/mailroom.py